src/v/compat/raft_compat.h

```diff
@@ -0,0 +1,246 @@
+    }
+
+    static void
+    to_json(raft::timeout_now_reply obj, json::Writer<json::StringBuffer>& wr) {
+        json_write(target_node_id);
+        json_write(term);
+        json_write(result);
+    }
+
+    static raft::timeout_now_reply from_json(json::Value& rd) {
+        raft::timeout_now_reply obj;
+        json_read(target_node_id);
+        json_read(term);
+        auto result = json::read_member_enum(rd, "result", obj.result);
+        switch (result) {
+        case 0:
+            obj.result = raft::timeout_now_reply::status::success;
+            break;
+        case 1:
+            obj.result = raft::timeout_now_reply::status::failure;
+            break;
+        default:
+            vassert(false, "invalid status: {}", result);
+        }
+        return obj;
+    }
+
+    static std::vector<compat_binary> to_binary(raft::timeout_now_reply obj) {
+        return compat_binary::serde_and_adl(obj);
+    }
+
+    static bool check(raft::timeout_now_reply obj, compat_binary test) {
+        return verify_adl_or_serde(obj, std::move(test));
+    }
+};
+
+/*
+ * raft::transfer_leadership_request
+ */
+template<>
+struct compat_check<raft::transfer_leadership_request> {
+    static constexpr std::string_view name
+      = "raft::transfer_leadership_request";
+
+    static std::vector<raft::transfer_leadership_request> create_test_cases() {
+        return generate_instances<raft::transfer_leadership_request>();
+    }
+
+    static void to_json(
+      raft::transfer_leadership_request obj,
+      json::Writer<json::StringBuffer>& wr) {
+        json_write(group);
+        json_write(target);
+    }
+
+    static raft::transfer_leadership_request from_json(json::Value& rd) {
+        raft::transfer_leadership_request obj;
+        json_read(group);
+        json_read(target);
+        return obj;
+    }
+
+    static std::vector<compat_binary>
+    to_binary(raft::transfer_leadership_request obj) {
+        return compat_binary::serde_and_adl(obj);
+    }
+
+    static bool
+    check(raft::transfer_leadership_request obj, compat_binary test) {
+        return verify_adl_or_serde(obj, std::move(test));
+    }
+};
+
+/*
+ * raft::transfer_leadership_reply
+ */
+template<>
+struct compat_check<raft::transfer_leadership_reply> {
+    static constexpr std::string_view name = "raft::transfer_leadership_reply";
+
+    static std::vector<raft::transfer_leadership_reply> create_test_cases() {
+        return generate_instances<raft::transfer_leadership_reply>();
+    }
+
+    static void to_json(
+      raft::transfer_leadership_reply obj,
+      json::Writer<json::StringBuffer>& wr) {
+        json_write(success);
+        json_write(result);
+    }
+
+    static raft::transfer_leadership_reply from_json(json::Value& rd) {
+        raft::transfer_leadership_reply obj;
+        json_read(success);
+        auto result = json::read_member_enum(rd, "result", obj.result);
+        switch (result) {
```

Comment:
Is there a reason to not just cast the result and assert that it's in the range of valid values?

Reply:
being in range doesn't mean that it matches an exact enum

Reply:
perhaps we should serialize the name of the enum and its value to preserve maximum semantics. i think this is something we can add as go along. on that point, i added a version field to the json test file.

Reply:
> merely casting to an invalid value is UB

checking that it's in range before casting should prevent this.

writing

```
  vassert(result >= static_cast<std::underlying_type_t<raft::errc>>(raft::errc::success) &&
               result <= static_cast<std::underlying_type_t<raft:errc>>(raft::errc::replicate_first_stage_exception));
  obj.result = raft::errc{result};
```

should do the same as manually listing all the cases.

Reply:
nothing prevents an enum from skipping values, right?

Reply:
i mean i guess the question is related to what the spec says exactly about when ub occurs

Reply:
If this works maybe it can be a utility within compat so all types can use it

Reply:
> nothing prevents an enum from skipping values, right?

in general, yes. In most redpanda cases this is not the case

> i mean i guess the question is related to what the spec says exactly about when ub occurs

I think what you said "casting to an invalid value is UB" is correct. But since we check that it's in the range of valid values, this should not be UB from my understanding.

Reply:
> vassert(result >= static_cast<std::underlying_type_t<raft::errc>>(raft::errc::success) && result <= static_cast<std::underlying_type_t<raft:errc>>(raft::errc::replicate_first_stage_exception)); obj.result = raft::errc{result};

This is how I did it in my `id_allocator` types patch (didn't realize you had a helper here).. I think you are right that it can result in an UB per spec...but it seems to work ok when using skipped values..

Reply:
i think it _might_ be UB, but it seems like a reasonable thing to do to check range and evaluate based on numeric value. I'll switch this in another patch. it'll make things simpler.